onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -43,14 +43,12 @@ static common::Status AllocateBufferUsingDeviceAllocatorFromShapeAndType(const T
   p_data = nullptr;
   if (shape_size > 0) {
     SafeInt<size_t> mem_size = 0;
-    if (!alloc->CalcMemSizeForArray(SafeInt<size_t>(shape_size), type->Size(), &mem_size))
+
+    if (!alloc->CalcMemSizeForArray(SafeInt<size_t>(shape_size), type->Size(), &mem_size)) {
       return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed memory size calculation");
+    }
 
-    if (alloc->Info().alloc_type == OrtArenaAllocator)
-      // Reserve() uses the device allocator to make the allocation
-      p_data = static_cast<IArenaAllocator*>(alloc.get())->Reserve(mem_size);
-    else
-      p_data = alloc->Alloc(mem_size);
+    p_data = alloc->Reserve(mem_size);
```

Comment:
Can just call Reserve() now as it is part of the IAllocator interface (and calls Alloc() by default). BFCArena provides a specialized override for Reserve()